Refactor mouse and keyboard handling in the web process to be frame specific

charliewolfe opened the pull request:
#### b1c1773aa4f7b1b065d8e262b0c6bb6d5c173b70
<pre>
Refactor mouse and keyboard handling in the web process to be frame specific
<a href="https://bugs.webkit.org/show_bug.cgi?id=260356">https://bugs.webkit.org/show_bug.cgi?id=260356</a>
rdar://114035729

Reviewed by Wenson Hsieh and Alex Christensen.

To implement mouse and keyboard events for site isolated frames, we need to be able to send
events to frames which are in a different process than the main frame. This patch works
towards this by refactoring how these events are handled in the web process to be able to
specify a frame. There still needs to be changes so that the events queued in the UI process
have support to be to sent to specified frames. For now, these events will all still go to
the main frame.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::LocalFrame):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::userInputBridge): Deleted.
* Source/WebCore/replay/UserInputBridge.cpp:
(WebCore::UserInputBridge::UserInputBridge):
(WebCore::UserInputBridge::handleContextMenuEvent):
(WebCore::UserInputBridge::handleMousePressEvent):
(WebCore::UserInputBridge::handleMouseReleaseEvent):
(WebCore::UserInputBridge::handleMouseMoveEvent):
(WebCore::UserInputBridge::handleMouseMoveOnScrollbarEvent):
(WebCore::UserInputBridge::handleMouseForceEvent):
(WebCore::UserInputBridge::handleKeyEvent):
(WebCore::UserInputBridge::handleAccessKeyEvent):
(WebCore::UserInputBridge::handleWheelEvent):
(WebCore::UserInputBridge::focusSetActive):
(WebCore::UserInputBridge::focusSetFocused):
(WebCore::UserInputBridge::scrollRecursively):
(WebCore::UserInputBridge::logicalScrollRecursively):
(WebCore::UserInputBridge::loadRequest):
(WebCore::UserInputBridge::reloadFrame):
(WebCore::UserInputBridge::stopLoadingFrame):
(WebCore::UserInputBridge::tryClosePage):
* Source/WebCore/replay/UserInputBridge.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::sendWheelEvent):
(WebKit::WebPageProxy::handleKeyboardEvent):
(WebKit::WebPageProxy::didReceiveEvent):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyContextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::dispatchWheelEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::isContextClick):
(WebKit::WebFrame::handleContextMenuEvent):
(WebKit::WebFrame::handleMouseEvent):
(WebKit::WebFrame::handleKeyEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::tryClose):
(WebKit::WebPage::loadRequest):
(WebKit::WebPage::stopLoading):
(WebKit::WebPage::reload):
(WebKit::WebPage::contextMenuAtPointInWindow):
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::handleWheelEvent):
(WebKit::WebPage::wheelEvent):
(WebKit::WebPage::keyEvent):
(WebKit::WebPage::scroll):
(WebKit::WebPage::logicalScroll):
(WebKit::WebPage::simulateMouseDown):
(WebKit::WebPage::simulateMouseUp):
(WebKit::WebPage::simulateMouseMotion):
(WebKit::isContextClick): Deleted.
(WebKit::handleContextMenuEvent): Deleted.
(WebKit::handleMouseEvent): Deleted.
(WebKit::handleKeyEvent): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/267124@main">https://commits.webkit.org/267124@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d47fd1e587f7dc7ecb04138641f8e99b7c5716b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16008 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17456 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14732 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15889 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18994 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16108 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15886 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16378 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13379 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18204 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14175 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21100 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14635 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17602 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12673 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14183 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3758 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18550 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14749 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebCore/replay/UserInputBridge.cpp:
```diff
@@ -41,123 +41,129 @@
 bool UserInputBridge::handleKeyEvent(const PlatformKeyboardEvent& keyEvent, InputSource)
 {
-    return Ref(m_page.focusController().focusedOrMainFrame())->eventHandler().keyEvent(keyEvent);
+    auto* page = m_frame.page();
```
It might make sense to not take in the frame here since we do `focusedOrMainFrame` and might even do that with site isolated key events.

Or maybe we can calculate `focusedOrMainFrame` in the UI process before sending the message to have the right frame here.

charliewolfe replied:
It would seem better if we could make sure this is always being called on the correct frame instead of checking in this function, I can look at how I could do that.

We will still need to make sure `focusedOrMainFrame()` works correctly for site isolation, I didn't try to fix that in this patch.

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3288,7 +3291,10 @@ void WebPageProxy::flushPendingMouseEventCallbacks()
 {
-    sendWithAsyncReply(Messages::WebPage::DispatchWheelEventWithoutScrolling(event), WTFMove(completionHandler));
+    if (!m_mainFrame)
+        return;
```
You want to call the `completionHandler` in this path too

aprotyas reviewed (commented):
Just a drive-by comment, not a real review

whsieh commented on Source/WebCore/replay/UserInputBridge.cpp:
```diff
@@ -41,123 +41,117 @@
 bool UserInputBridge::scrollRecursively(ScrollDirection direction, ScrollGranularity granularity, InputSource)
 {
-    return Ref(m_page.focusController().focusedOrMainFrame())->eventHandler().scrollRecursively(direction, granularity, nullptr);
+    return m_frame.eventHandler().scrollRecursively(direction, granularity, nullptr);
```
Nit - it's probably a good idea to continue protecting the `frame`, since these methods are capable of triggering layout (and can therefore trigger frame destruction, unless otherwise protected).

whsieh commented on Source/WebCore/replay/UserInputBridge.cpp:
```diff
@@ -41,123 +41,117 @@
 bool UserInputBridge::logicalScrollRecursively(ScrollLogicalDirection direction, ScrollGranularity granularity, InputSource)
 {
-    return Ref(m_page.focusController().focusedOrMainFrame())->eventHandler().logicalScrollRecursively(direction, granularity, nullptr);
+    return m_frame.eventHandler().logicalScrollRecursively(direction, granularity, nullptr);
```
(Ditto)

whsieh commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -47,8 +47,8 @@ messages -> WebPage LegacyReceiver {
-    KeyEvent(WebKit::WebKeyboardEvent event)
-    MouseEvent(WebKit::WebMouseEvent event, std::optional<Vector<WebKit::SandboxExtension::Handle>> sandboxExtensions)
+    KeyEvent(std::optional<WebCore::FrameIdentifier> frameID, WebKit::WebKeyboardEvent event)
+    MouseEvent(std::optional<WebCore::FrameIdentifier> frameID, WebKit::WebMouseEvent event, std::optional<Vector<WebKit::SandboxExtension::Handle>> sandboxExtensions)
```
Nit - I think it might be a bit cleaner to always take a `FrameIdentifier` instead of a `std::optional<FrameIdentifier>`, and using `std::optional` to mean "main frame". I _suspect_ that in all cases where we're able to get the main frame's ID in the web process when the `std::optional<FrameIdentifier>` is `nullopt`, the UI process should already be aware of the main frame's ID. (That said, I could be missing a subtlety here!)

charliewolfe replied:
I had this initially, but I changed it because there was cases where the main frame didn’t exist, but we still expected didReceiveEvent to be called from the web process to indicate that the event was not handled.

TestWebKitAPI.WebKit.DidNotHandleKeyDown was the relevant test.

I can look to see if there is a better way to handle that situation.

whsieh replied:
I see, that's indeed a bit subtle!

Perhaps a simpler alternative would be to just skip the round trip and consider the event unhandled in the UI process if there's no frame that's capable of handling the event — that said, I think the `std::optional<>` should be okay too (maybe with a brief comment explaining when the ID is `std::nullopt`).

whsieh reviewed (commented):
This looks good! I left a few relatively minor comments (and am also waiting a bit to see if others have feedback as well).

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1940,7 +1940,7 @@ void WebPage::loadRequest(LoadParameters&& loadParameters)
 
     if (frame.get() == m_mainFrame.ptr())
-        corePage()->userInputBridge().loadRequest(WTFMove(frameLoadRequest));
+        m_mainFrame->coreLocalFrame()->userInputBridge().loadRequest(WTFMove(frameLoadRequest));
```
We may want to null check all these new coreLocalFrame() calls to keep things from crashing while we're developing site isolation.

charliewolfe committed:
Refactor mouse and keyboard handling in the web process to be frame specific
https://bugs.webkit.org/show_bug.cgi?id=260356
rdar://114035729

Reviewed by Wenson Hsieh and Alex Christensen.

To implement mouse and keyboard events for site isolated frames, we need to be able to send
events to frames which are in a different process than the main frame. This patch works
towards this by refactoring how these events are handled in the web process to be able to
specify a frame. There still needs to be changes so that the events queued in the UI process
have support to be to sent to specified frames. For now, these events will all still go to
the main frame.

* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::LocalFrame):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::Page):
* Source/WebCore/page/Page.h:
(WebCore::Page::userInputBridge): Deleted.
* Source/WebCore/replay/UserInputBridge.cpp:
(WebCore::UserInputBridge::UserInputBridge):
(WebCore::UserInputBridge::handleContextMenuEvent):
(WebCore::UserInputBridge::handleMousePressEvent):
(WebCore::UserInputBridge::handleMouseReleaseEvent):
(WebCore::UserInputBridge::handleMouseMoveEvent):
(WebCore::UserInputBridge::handleMouseMoveOnScrollbarEvent):
(WebCore::UserInputBridge::handleMouseForceEvent):
(WebCore::UserInputBridge::handleKeyEvent):
(WebCore::UserInputBridge::handleAccessKeyEvent):
(WebCore::UserInputBridge::handleWheelEvent):
(WebCore::UserInputBridge::focusSetActive):
(WebCore::UserInputBridge::focusSetFocused):
(WebCore::UserInputBridge::scrollRecursively):
(WebCore::UserInputBridge::logicalScrollRecursively):
(WebCore::UserInputBridge::loadRequest):
(WebCore::UserInputBridge::reloadFrame):
(WebCore::UserInputBridge::stopLoadingFrame):
(WebCore::UserInputBridge::tryClosePage):
* Source/WebCore/replay/UserInputBridge.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::sendWheelEvent):
(WebKit::WebPageProxy::handleKeyboardEvent):
(WebKit::WebPageProxy::didReceiveEvent):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageCopyContextMenuAtPointInWindow):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::dispatchWheelEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::isContextClick):
(WebKit::WebFrame::handleContextMenuEvent):
(WebKit::WebFrame::handleMouseEvent):
(WebKit::WebFrame::handleKeyEvent):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::tryClose):
(WebKit::WebPage::loadRequest):
(WebKit::WebPage::stopLoading):
(WebKit::WebPage::reload):
(WebKit::WebPage::contextMenuAtPointInWindow):
(WebKit::WebPage::mouseEvent):
(WebKit::WebPage::handleWheelEvent):
(WebKit::WebPage::wheelEvent):
(WebKit::WebPage::keyEvent):
(WebKit::WebPage::scroll):
(WebKit::WebPage::logicalScroll):
(WebKit::WebPage::simulateMouseDown):
(WebKit::WebPage::simulateMouseUp):
(WebKit::WebPage::simulateMouseMotion):
(WebKit::isContextClick): Deleted.
(WebKit::handleContextMenuEvent): Deleted.
(WebKit::handleMouseEvent): Deleted.
(WebKit::handleKeyEvent): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/267124@main

fujii commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3248,6 +3248,9 @@ void WebPageProxy::processNextQueuedMouseEvent()
         return;
 
+    if (!m_mainFrame)
+        return;
```
I'm observing a problem for Windows port.
If I move a mouse on a window immediately after MiniBrowser is started, all mouse events are ignored.
`WebPageProxy::processNextQueuedMouseEvent()` returns early because !m_mainFrame, then `WebPageProxy::processNextQueuedMouseEvent()` never be called.

fujii replied:
I confirmed this patch fixes the problem. 
~~~diff
diff --git a/Source/WebKit/UIProcess/WebPageProxy.cpp b/Source/WebKit/UIProcess/WebPageProxy.cpp
index e67f06539ee6..df4aefb34179 100644
--- a/Source/WebKit/UIProcess/WebPageProxy.cpp
+++ b/Source/WebKit/UIProcess/WebPageProxy.cpp
@@ -3212,6 +3212,9 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
     if (!hasRunningProcess())
         return;
 
+    if (!m_mainFrame)
+        return;
+
 #if ENABLE(CONTEXT_MENU_EVENT)
     if (event.button() == WebMouseEventButton::RightButton && event.type() == WebEventType::MouseDown) {
         ASSERT(m_contextMenuPreventionState != EventPreventionState::Waiting);
~~~

fujii commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -3082,21 +3086,23 @@ WebContextMenu& WebPage::contextMenu()
-    if (!localMainFrame)
-        return nullptr;
-    Ref mainFrame = *localMainFrame;
-    bool handled = corePage()->userInputBridge().handleContextMenuEvent(mousePressEvent, mainFrame);
```
I'm observing another problem for Windows port.
Layout tests using a context menu are timing out.
The original code didn't show a context menu for contextMenuAtPointInWindow.
I attached my suggested patch. https://bugs.webkit.org/show_bug.cgi?id=260541#c2
